### PR TITLE
Support ChainerX in `F.copy`

### PR DESCRIPTION
**Pull request by niboshi**
ChainerX support in `F.copy`.

(Required in #6857, separated from #6859)

~TODO: test~

### CODE REVIEW

**Comment by hvy** [comment]
A drop by comment but should we not implement `forward_chainerx` instead?

File: chainer/functions/array/copy.py [hunk 1]
```diff
@@ -36,24 +41,38 @@ def copy(x, dst):
...
     """
-    y, = Copy(dst).apply((x,))
+    in_device = backend.get_device_from_array(
+        x.array if isinstance(x, chainer.Variable) else x)
+    out_device = chainer.get_device(dst)
+
+    is_chainerx = in_device.xp is chainerx
+    if is_chainerx != (out_device.xp is chainerx):
+        raise RuntimeError(
+            'F.copy does not support a copy between a non-ChainerX device and '
```
Comment (hvy) [inline chainer/functions/array/copy.py]: A major nitpick but I think the following is a more common way to put it.
```suggestion
            'F.copy does not support copies between non-ChainerX devices and '
```

File: chainer/functions/array/copy.py [hunk 2]
```diff
@@ -36,24 +41,38 @@ def copy(x, dst):
...
+            'a ChainerX device.\n'
```
Comment (hvy) [inline chainer/functions/array/copy.py]: Following the comment above.
```suggestion
            'ChainerX devices.\n'
```

File: chainer/functions/array/copy.py [hunk 3]
```diff
@@ -36,24 +41,38 @@ def copy(x, dst):
...
+        >>> x = chainer.Variable(x_arr)
+        >>> x.device
+        '@numpy'
+        >>> y = F.copy(x, '@cupy:0') # from CPU (NumPy) to GPU 0 (CuPy)
+        >>> y.device
+        '@cupy:0'
+
+    .. note::
+        Copies between non-ChainerX devices and ChainerX devices are not
+        supported.
```
Comment (hvy) [inline chainer/functions/array/copy.py]: Could you explain why it's not supported (maybe not in the documentation but in this PR)?

Comment (niboshi) [reply chainer/functions/array/copy.py]: Because it would be extremely difficult to support backprop.

File: chainer/functions/array/copy.py (hunk 1 above)
Comment (niboshi) [reply chainer/functions/array/copy.py]: Thanks, fixed.

File: chainer/functions/array/copy.py (hunk 2 above)
Comment (niboshi) [reply chainer/functions/array/copy.py]: Fixed.

File: tests/chainer_tests/functions_tests/array_tests/test_copy.py [hunk 4]
```diff
@@ -18,97 +18,184 @@ def _to_gpu(x, device_id):
...
+        assert backend.get_device_from_array(x_grad) == src_device
+        numpy.testing.assert_array_equal(_numpy_device.send(x_grad), self.gy)
+
+    def test_double_backward(self, src_backend_config, dst_backend_config):
+        x = src_backend_config.get_array(self.x)
+        gy = dst_backend_config.get_array(self.gy)
+        ggx = src_backend_config.get_array(self.ggx)
+        dst_device = dst_backend_config.device
+
+        x_var = chainer.Variable(x, requires_grad=True)
```
Comment (hvy) [inline tests/chainer_tests/functions_tests/array_tests/test_copy.py]: Do we need the `requires_grad=True`? I think it increases the reader's load a bit since there's no such argument passed in `test_backward`.

File: tests/chainer_tests/functions_tests/array_tests/test_copy.py [hunk 5]
```diff
@@ -18,97 +18,184 @@ def _to_gpu(x, device_id):
...
+
+        y_var = functions.copy(x_var, dst_device)
 
-        self.assertEqual(self.x_data.dtype, self.dtype)
-        numpy.testing.assert_array_equal(self.x_data, cuda.to_cpu(y.data))
+        # TODO(niboshi): Remove this workround after Variable.grad.setter is
+        # fixed so that it calls gy.require_grad() internally.
```
Comment (hvy) [inline tests/chainer_tests/functions_tests/array_tests/test_copy.py]: Can we address this TODO now by constructing a `Variable` gradient and call the `grad_var` setter instead (and preferably similarly in other gradient assignments in this file for consistency)?

Comment (niboshi) [reply tests/chainer_tests/functions_tests/array_tests/test_copy.py]: What's the point of your suggestion?
I think both ways are equally correct.

File: tests/chainer_tests/functions_tests/array_tests/test_copy.py (hunk 4 above)
Comment (niboshi) [reply tests/chainer_tests/functions_tests/array_tests/test_copy.py]: `requires_grad=True` is essential in this test and it's better to make it explicit.

File: chainer/functions/array/copy.py (hunk 3 above)
Comment (hvy) [reply chainer/functions/array/copy.py]: I see, it doesn't get along very well with the `forward_chainerx` forward implementation separation and the fallback mechanism I guess.

File: tests/chainer_tests/functions_tests/array_tests/test_copy.py (hunk 5 above)
Comment (hvy) [reply tests/chainer_tests/functions_tests/array_tests/test_copy.py]: I think my suggestion is simpler since it's xp-agnostic but then I also meant we could remove the TODO comment since it's not a temporary workaround.

File: tests/chainer_tests/functions_tests/array_tests/test_copy.py (hunk 4 above)
Comment (hvy) [reply tests/chainer_tests/functions_tests/array_tests/test_copy.py]: Is it not the case for `test_backward` ?

File: tests/chainer_tests/functions_tests/array_tests/test_copy.py (hunk 5 above)
Comment (niboshi) [reply tests/chainer_tests/functions_tests/array_tests/test_copy.py]: I'm not persuaded.
Both ways are xp-agnostic and both are equally correct as I said.
It shouldn't matter whether we need TODO, because it doesn't mean it's a wrong way.

File: tests/chainer_tests/functions_tests/array_tests/test_copy.py (hunk 4 above)
Comment (niboshi) [reply tests/chainer_tests/functions_tests/array_tests/test_copy.py]: It is.

File: tests/chainer_tests/functions_tests/array_tests/test_copy.py (hunk 5 above)
Comment (hvy) [reply tests/chainer_tests/functions_tests/array_tests/test_copy.py]: Okay, let's leave it as is since the TODO will be addressed most likely and in that case `.grad` will be simpler.

**Comment by hvy** [comment]
Could you take a look at the Travis failures?

**Comment by niboshi** [comment]
#7145 is needed for `TestCopyCudaDummyDevice`.
(the behavior of `F.copy` is not changed in this PR).

**Comment by niboshi** [comment]
Instead of relying on #7145, I fixed `F.copy` to handle `DummyDevice` in its own.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
LGTM.

**Comment by hvy** [comment]
Jenkins, test this please.
